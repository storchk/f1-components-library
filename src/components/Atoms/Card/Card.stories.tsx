import type { Meta, StoryObj } from '@storybook/react'

import { Card } from './Card'
import type { CardProps } from './Card.types'
import { Typography } from '../Typography'
import { Badge } from '../Badge'

export default {
  title: 'Atoms/Card',
  component: Card,
} as Meta

type Story = StoryObj<CardProps>

export const Primary: Story = {
  render: props => (
    <Card variant="primary" {...props}>
      <div>
        <Typography>
          Relive the frantic Sprint Shootout as Verstappen takes P1 and Hamilton exits early
        </Typography>
        <Typography>
          Relive the frantic Sprint Shootout as Verstappen takes P1 and Hamilton exits early
        </Typography>
        <Typography>
          Relive the frantic Sprint Shootout as Verstappen takes P1 and Hamilton exits early
        </Typography>
      </div>
    </Card>
  ),
}
export const Secondary: Story = {
  render: props => (
    <Card variant="secondary" {...props}>
      <Badge label="Highlight" />
      <div>
        <Typography>
          Relive the frantic Sprint Shootout as Verstappen takes P1 and Hamilton exits early
        </Typography>
        <Typography>
          Relive the frantic Sprint Shootout as Verstappen takes P1 and Hamilton exits early
        </Typography>
      </div>
    </Card>
  ),
}
