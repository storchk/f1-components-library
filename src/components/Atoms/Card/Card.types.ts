import type { ThemeType } from '../../../theme'

export type CardVariant = 'primary' | 'secondary'

export type CardProps = {
  variant?: CardVariant
  children: string | JSX.Element | JSX.Element[]
  $noPadding?: boolean
}

export type StyledCardProps = {
  $variant: CardVariant
} & Omit<CardProps, 'variant'>

export type GetCardVariantsProps = {
  theme: ThemeType
} & Pick<StyledCardProps, '$variant'>
