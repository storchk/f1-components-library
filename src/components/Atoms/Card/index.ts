export { Card } from './Card'
export type { CardProps, CardVariant } from './Card.types'
