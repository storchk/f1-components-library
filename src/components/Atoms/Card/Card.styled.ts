import styled, { css } from 'styled-components'

import type { StyledCardProps } from './Card.types'
import { getCardVariants } from './Card.variants'
import { StyledTypography } from '../Typography'

export const StyledCardText = styled(StyledTypography)``

export const StyledCard = styled.div<StyledCardProps>(
  ({ theme, $variant }) => css`
    display: block;
    transition: all 0.5s cubic-bezier(0.2, 0, 0.05, 1);

    ${getCardVariants({ theme, $variant })}

    &:hover {
      cursor: pointer;
    }
    > div {
      padding: ${theme.spacing.xl} ${theme.spacing.sm};
    }
  `
)
