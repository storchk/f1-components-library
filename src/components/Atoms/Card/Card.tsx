import { StyledCard, StyledCardText } from './Card.styled'
import type { CardProps } from './Card.types'
export const Card = ({ variant = 'primary', children, ...props }: CardProps): JSX.Element => {
  return (
    <StyledCard $variant={variant} {...props}>
      <div>
        <StyledCardText>{children}</StyledCardText>
      </div>
    </StyledCard>
  )
}
