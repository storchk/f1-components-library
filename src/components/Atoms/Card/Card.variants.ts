import { css } from 'styled-components'

import type { GetCardVariantsProps } from './Card.types'
import { StyledCardText } from './Card.styled'

export const getCardVariants = ({ theme, $variant }: GetCardVariantsProps) => {
  switch ($variant) {
    case 'primary':
    default:
      return css`
        border-bottom: solid ${theme.border.width.small};
        border-right: solid ${theme.border.width.small};
        border-bottom-right-radius: ${theme.spacing.md};
        background-color: ${theme.colors.pureWhite};
        border-bottom-color: ${theme.colors.contrast2};
        border-right-color: ${theme.colors.contrast2};

        &:hover {
          border-bottom-color: ${theme.colors.primary};
          border-right-color: ${theme.colors.primary};
        }
      `
    case 'secondary':
      return css`
        background-color: ${theme.colors.pureWhite};
        border-bottom-left-radius: ${theme.spacing.md};
        border-bottom-right-radius: ${theme.spacing.md};

        &:hover {
          background-color: ${theme.colors.secondary};

          ${StyledCardText} {
            color: ${theme.colors.pureWhite};
          }
        }
      `
  }
}
