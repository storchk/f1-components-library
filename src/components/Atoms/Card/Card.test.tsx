import { testing } from '../../../testing'
import { Card } from './Card'
import { CardProps } from './Card.types'
import { axe } from 'jest-axe'

const { render, screen, renderer } = testing
const mockText = 'Hello world! I am using React'

const renderComponent = (props?: Partial<CardProps>) => render(<Card {...props}>{mockText}</Card>)
describe('Atoms: Card', () => {
  it('should render button with text', () => {
    renderComponent()
    const text = screen.getByText(/Hello world! I am using React/i)
    expect(text).toBeInTheDocument()
  })

  it('snapshot | variant: primary', () => {
    const view = renderer(<Card variant="primary">{mockText}</Card>)
    expect(view).toMatchSnapshot()
  })
  it('snapshot | variant: secondary', () => {
    const view = renderer(<Card variant="secondary">{mockText}</Card>)
    expect(view).toMatchSnapshot()
  })

  test('A11y - No violations detected', async () => {
    const { container } = renderComponent()
    expect(await axe(container)).toHaveNoViolations()
  })
})
