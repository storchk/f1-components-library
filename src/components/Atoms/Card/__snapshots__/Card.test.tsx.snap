// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atoms: Card snapshot | variant: primary 1`] = `
.c1 {
  font-size: inherit;
  font-weight: 400;
  color: inherit;
  font-family: Titillium Web;
}

.c0 {
  display: block;
  -webkit-transition: all 0.5s cubic-bezier(0.2,0,0.05,1);
  transition: all 0.5s cubic-bezier(0.2,0,0.05,1);
  border-bottom: solid 2px;
  border-right: solid 2px;
  border-bottom-right-radius: 24px;
  background-color: #fff;
  border-bottom-color: #f5f2ef;
  border-right-color: #f5f2ef;
}

.c0:hover {
  border-bottom-color: #e10600;
  border-right-color: #e10600;
}

.c0:hover {
  cursor: pointer;
}

.c0 > div {
  padding: 40px 16px;
}

<div
  className="c0"
>
  <div>
    <div
      className="c1"
    >
      Hello world! I am using React
    </div>
  </div>
</div>
`;

exports[`Atoms: Card snapshot | variant: secondary 1`] = `
.c2 {
  font-size: inherit;
  font-weight: 400;
  color: inherit;
  font-family: Titillium Web;
}

.c0 {
  display: block;
  -webkit-transition: all 0.5s cubic-bezier(0.2,0,0.05,1);
  transition: all 0.5s cubic-bezier(0.2,0,0.05,1);
  background-color: #fff;
  border-bottom-left-radius: 24px;
  border-bottom-right-radius: 24px;
}

.c0:hover {
  background-color: #1f1f27;
}

.c0:hover .c1 {
  color: #fff;
}

.c0:hover {
  cursor: pointer;
}

.c0 > div {
  padding: 40px 16px;
}

<div
  className="c0"
>
  <div>
    <div
      className="c1 c2"
    >
      Hello world! I am using React
    </div>
  </div>
</div>
`;
